subt/ros/robot/src/filter_pointcloud.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/python
+"""
+  Filter PointCloud2 received from X4 drone for Octomap server
+  - remove nearby propellers (up to 31cm) in the RGBD camera view
+  - replace xyz = (inf, inf, inf) to range larger than 10m
+"""
+# http://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/PointCloud2.html
+# http://docs.ros.org/en/api/sensor_msgs/html/point__cloud2_8py_source.html
+import ctypes
+import struct
+
+import rospy
+from sensor_msgs.msg import PointCloud2, PointField
+
+import numpy as np
+
+
+class FilterPointCloud:
+    def __init__(self):
+#        self.points_subscriber = rospy.Subscriber("/input_points", PointCloud2, self.points_callback)
+        self.points_subscriber = rospy.Subscriber("/A100LXM/rgbd_camera/depth/points", PointCloud2, self.points_callback)
+#        self.points_publisher = rospy.Publisher("/output_points", PointCloud2, queue_size = 1)
+        self.points_publisher = rospy.Publisher("/points_cleaned", PointCloud2, queue_size=1)
+
+        # create flat "infinity" background
+        size = 640 * 480
+        self.inf_x = np.zeros(size, dtype=np.float32)
+        self.inf_y = np.zeros(size, dtype=np.float32)
+        self.inf_z = np.zeros(size, dtype=np.float32)
+        fx = 554.25469
+        MAX_VALUE = 11.0
+        for x in range(640):
+            for y in range(480):
+                pos = y * 640 + x
+                self.inf_x[pos] = MAX_VALUE
+                self.inf_y[pos] = MAX_VALUE * (320.5 - x)/fx
+                self.inf_z[pos] = MAX_VALUE * (240.5 - y)/fx
+
+    def points_callback(self, msg):
+        assert msg.height == 480, msg.height
+        assert msg.width == 640, msg.width
+        assert msg.point_step == 24, msg.point_step
+        assert msg.row_step == 640 * 24, msg.row_step
+        arr = np.frombuffer(msg.data, dtype=np.float32)
+        x = arr[::6].copy()
+        y = arr[1::6].copy()
+        z = arr[2::6].copy()
```

Comment:
Can we try something else?

1. inf_x, inf_y and inf_z can be a single 480x640x3 `background` array instead. Or a 307200x3 array. See my comment about depth.py above. Let's go with the 480x640x3 version for simplicity.
2. data = np.frombuffer(msg.data, dtype=np.float32).reshape((480, 640, 6))
3. xyz = data[:,:,:3]
3. mask = np.logical_or(np.isposinf(xyz[:,:,0]), xyz[:,:,0] < 0.32)
4. xyz[:,:,:] = np.where(mask, background, xyz)
5. new_data = data.tobytes()
6. new_msg = PointCloud2(..., data=new_data)

Does this work? (I have not tried to validate it.)

Reply:
I will check, but not now (timeout unfortunately). As I said first I need some way how to make sure it is still working and running the simulation is not the way ... (ROS + Python2, with launch of tens of nodes ... sigh.
But thanks a lot, I will come back to this (I will have to) with various robot types.

Reply:
I originally supposed to create extra sub-branch to exercise this `numpy` parts, but if I will commit intermediate results you may comment if I am going wrong direction ;) ... so WIP at the moment.